### PR TITLE
Add "disasm sym_info.py" script

### PR DESCRIPTION
**Pull request by Dragorn421**
This is useful for investigating the expected memory layout

~~**Bleeding edge spimdisasm is currently required** in order for the csv written by `saveContextToFile` to have file names info set~~
**requires spimdisasm >= 1.28.1** (just `make setup` and the venv will update it)

<details>

<summary>Some examples of running the script:</summary>

<details>

<summary>help</summary>

```
$ ./tools/disasm/sym_info.py --help
usage: sym_info.py [-h] [--version VERSION] [--around AROUND] [--range RANGE] [--file]
                   [--labels]
                   sym_or_vma

Print informations on a symbol/address (and possibly surrounding symbols) from the
spimdisasm disassembly context saved in expected/build/VERSION/context.csv

positional arguments:
  sym_or_vma

options:
  -h, --help            show this help message and exit
  --version VERSION, -v VERSION
                        oot version (default: ntsc-1.2)
  --around AROUND, -n AROUND
                        how many symbols to show around the target (at least)
  --range RANGE, -r RANGE
                        show symbols within this range from the target (at least)
  --file, -f            show symbols within the file section of the target (at least)
  --labels, -l          also show label symbols
```

</details>

<details>

<summary>look up a single symbol by name</summary>

```
$ ./tools/disasm/sym_info.py EnDaiku_Update
EnDaiku_Update = 0x80A8DE90
src/overlays/actors/ovl_En_Daiku/z_en_daiku .text
    ...
    EnDaiku_Update 0x80A8DE90 ROM:4480 (@function) (sz=0x118)
    ...
```

the ellipsis means there are symbols before and after in the file

</details>

<details>

<summary>look up symbol(s) by address</summary>

```
$ ./tools/disasm/sym_info.py 0x80A8DE94
None = 0x80A8DE94
src/overlays/actors/ovl_En_Daiku/z_en_daiku .text
    ...
    EnDaiku_Update 0x80A8DE90 ROM:4480 (@function) (sz=0x118)
    EnDaiku_Draw 0x80A8DFA8 ROM:4760 (@function) (sz=0x128)
    ...
```

</details>

<details>

<summary>look up symbols within 0x300 bytes (after or before) of 0x80A8E000</summary>

```
$ ./tools/disasm/sym_info.py 0x80A8E000 --range 0x300
None = 0x80A8E000
src/overlays/actors/ovl_En_Daiku/z_en_daiku .text
    ...
    EnDaiku_Update 0x80A8DE90 ROM:4480 (@function) (sz=0x118)
    EnDaiku_Draw 0x80A8DFA8 ROM:4760 (@function) (sz=0x128)
    EnDaiku_OverrideLimbDraw 0x80A8E0D0 ROM:5056 (@function) (sz=0x84)
    EnDaiku_PostLimbDraw 0x80A8E154 ROM:5188 (@function) (sz=0x7C)
src/overlays/actors/ovl_En_Daiku/z_en_daiku .data
    D_80A8E1D0_unknown 0x80A8E1D0 ROM:5312 (sz=0x4)
    D_80A8E1F0_unknown 0x80A8E1F0 ROM:5344 (sz=0x4)
    D_80A8E21C_unknown 0x80A8E21C ROM:5388 (sz=0x4)
    D_80A8E228_unknown 0x80A8E228 ROM:5400 (sz=0x4)
    D_80A8E248_unknown 0x80A8E248 ROM:5432 (sz=0x4)
    D_80A8E298_unknown 0x80A8E298 ROM:5512 (sz=0x4)
    D_80A8E2A4_unknown 0x80A8E2A4 ROM:5524 (sz=0x4)
    D_80A8E2D8_unknown 0x80A8E2D8 ROM:5576 (sz=0x4)
    D_80A8E2E4_unknown 0x80A8E2E4 ROM:5588 (sz=0x4)
    D_STR_80A8E2F4_unknown 0x80A8E2F4 ROM:5604 (sz=0x4)
src/overlays/actors/ovl_En_Daiku/z_en_daiku .rodata
    RO_FLT_80A8E300_unknown 0x80A8E300 ROM:5616 (f32) (sz=0x4)
    ...
```

note: --range can also be used with a symbol name (like all options)

</details>

<details>

<summary>look up a symbol by name and the 3 symbols before and after too</summary>

```
$ ./tools/disasm/sym_info.py EnDaiku_Draw --around 3
EnDaiku_Draw = 0x80A8DFA8
src/overlays/actors/ovl_En_Daiku/z_en_daiku .text
    ...
    EnDaiku_EscapeSuccess 0x80A8DB48 ROM:3640 (@function) (sz=0x13C)
    EnDaiku_EscapeRun 0x80A8DC84 ROM:3956 (@function) (sz=0x20C)
    EnDaiku_Update 0x80A8DE90 ROM:4480 (@function) (sz=0x118)
    EnDaiku_Draw 0x80A8DFA8 ROM:4760 (@function) (sz=0x128)
    EnDaiku_OverrideLimbDraw 0x80A8E0D0 ROM:5056 (@function) (sz=0x84)
    EnDaiku_PostLimbDraw 0x80A8E154 ROM:5188 (@function) (sz=0x7C)
src/overlays/actors/ovl_En_Daiku/z_en_daiku .data
    D_80A8E1D0_unknown 0x80A8E1D0 ROM:5312 (sz=0x4)
    ...
```

</details>

<details>

<summary>look up a symbol and also all symbols in its file-section</summary>

```
$ ./tools/disasm/sym_info.py bootproc -f
bootproc = 0x80000498
src/boot/boot_main .text
    bootclear 0x80000460 ROM:0 (@function) (sz=0x38)
    bootproc 0x80000498 ROM:56 (@function) (sz=0x108)
```

</details>

</details>

### CODE REVIEW

File: tools/disasm/sym_info.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+        type=int,
+        default=0,
+        help="how many symbols to show around the target (at least)",
+    )
+    parser.add_argument(
+        "--range",
+        "-r",
+        type=lambda v: int(v, 0),
+        default=0,
+        help="show symbols within this range from the target (at least)",
```
Comment (cadmic) [inline tools/disasm/sym_info.py:60]: "from the target" -> "after the target"?

File: tools/disasm/sym_info.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    )
+    parser.add_argument(
+        "--file",
+        "-f",
+        action="store_true",
+        help="show symbols within the file section of the target (at least)",
```
Comment (cadmic) [inline tools/disasm/sym_info.py:66]: I'm not sure what a "file section" is, maybe just spell out "symbols within the same file and same section as the target" or something?

File: tools/disasm/sym_info.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    )
+    parser.add_argument(
+        "--labels",
+        "-l",
+        action="store_true",
+        help="also show label symbols",
```
Comment (cadmic) [inline tools/disasm/sym_info.py:72]: clarify "branch and jump table labels"? (Actually are jump table labels included? looking at `LABELS_TYPES` above)

File: tools/disasm/sym_info.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
+#!/usr/bin/env python3
+# SPDX-FileCopyrightText: 2024 Dragorn421
```
Comment (cadmic) [inline tools/disasm/sym_info.py:2]: why not zeldaret?

File: tools/disasm/sym_info.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+                args.file
+                and (
+                    fs.vma_start
+                    <= target_vma
+                    < (
+                        filesections[i_fs + 1].vma_start
+                        if i_fs + 1 < len(filesections)
+                        else 0x1_0000_0000
+                    )
+                )
```
Comment (cadmic) [inline tools/disasm/sym_info.py:193]: might just be the formatting, but personally I'd rather read
`fs.vma_start <= target_vma and (i_fs + 1 < len(filesections) and target_vma < filesections[i_fs + 1].vma_start)` or something. Or factor out into a variable `vma_end`?

**Review by cadmic — COMMENTED** [review]
Cool, seems really useful
I'd like to wait for spimdisasm to get a release though

File: tools/disasm/sym_info.py (hunk 1 above)
Comment (Dragorn421) [reply tools/disasm/sym_info.py:60]: it also shows the symbols before within that range
for example, `./tools/disasm/sym_info.py 0x100 --range 0x80` would show symbols in the range 0x80 to 0x180

(am open to reworking how it works)

File: tools/disasm/sym_info.py (hunk 4 above)
Comment (Dragorn421) [reply tools/disasm/sym_info.py:2]: I copypasted my template and didn't think of it, will change

File: tools/disasm/sym_info.py (hunk 1 above)
Comment (cadmic) [reply tools/disasm/sym_info.py:60]: I think it's fine, just would like something similar to "around the target" then

File: tools/disasm/sym_info.py (hunk 3 above)
Comment (Dragorn421) [reply tools/disasm/sym_info.py:72]: jump tables labels should ideally be toggled on by this flag, I just didn't find their type to add to LABELS_TYPES

**Review by cadmic — APPROVED** [review]

**Review by fig02 — APPROVED** [review]

**Comment by Dragorn421** [comment]
spimdisasm got a release
I updated requirements.txt, and also made disasm.py check its version

**Review by cadmic — APPROVED** [review]

**Pull request merged**